From c9f1c04d9d6d4b9b893e49ddb409ba8954f83bf8 Mon Sep 17 00:00:00 2001
From: Johannes Thoma <johannes@johannesthoma.com>
Date: Thu, 25 Aug 2022 16:47:14 +0200
Subject: [PATCH] drbd: Check all connections for two phase commit support in
 do_change_disk_state()

Before this commit only the first connection was checked for supporting
two phase commit (protocol level 110 or higher). In some cornercases
when there were many detach/attach and connect/disconnect load happening
this could lead to a wrong (=false) return value of do_change_disk_state().
More precisely if the connection checked was up but no protocol level
was negotiated yet (agreed protocol level being min_level, usually 86)
a wrong cluster_wide_state_change supported flag was computed.

This lead to a situation where a local disk was attached without
notifying all the peers via a two phase commit cycle.
The peer, in turn, did not update the original node's peer
disk state (still in D_DISKLESS instead of D_NEGOTIATING)
and didn't send its state back which lead the
original node thinking (wrongly) that the replation state of
the peer is L_NEGOTIATING (it was set to off earlier in
finish_state_change() when this disk transisted locally from
not D_NEGOTIATING to D_NEGOTIATING). As a consequence, the
original node thought that negotiation hasn't finished and
the node's disk state was stuck in Negotiating forever.

Note that this issue only arised when there were more than
two diskful nodes.

This commit uses the existing supports_two_phase_commit()
function to check if two phase commit is supported on the
resource. This function checks first if a connection is
fully established and the looks on the agreed protocol
version.

This commit fixes the Negotiating stuckness we observed
when experimenting with WinDRBD and LINSTOR for Windows
(which - wrongly - does detach / attach on every drbdadm
adjust, this will be fixed in an extra commit).

Signed-off-by: Johannes Thoma <johannes@johannesthoma.com>
---
 drbd/drbd_state.c | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/drbd/drbd_state.c b/drbd/drbd_state.c
index e5b48572f..c75ae5597 100644
--- a/drbd/drbd_state.c
+++ b/drbd/drbd_state.c
@@ -5240,10 +5240,8 @@ static bool do_change_disk_state(struct change_context *context, enum change_pha
 	if (device->disk_state[NOW] == D_ATTACHING &&
 	    context->val.disk == D_NEGOTIATING) {
 		if (device_has_peer_devices_with_disk(device)) {
-			struct drbd_connection *connection =
-				first_connection(device->resource);
 			cluster_wide_state_change =
-				connection && connection->agreed_pro_version >= 110;
+				supports_two_phase_commit(device->resource);
 		} else {
 			/* very last part of attach */
 			context->val.disk = disk_state_from_md(device);
-- 
2.17.1

